### DISCUSSION TITLE
API to allow Backup Flowise Chatflows to Git or Allow Easy Undo/Revision Control?

### DISCUSSION THREAD

**Discussion by dkindlund** [FlowiseAI/Flowise#1585] [Ideas]
Hey @HenryHengZJ , just a small idea... as developers move their chatflows in Flowise to production, I realize that "Exporting Chatflows" isn't really going to work, long-term.  Specifically, it's currently very easy for developers to make a mistake in a Flowise chatflow and not easily "undo" or "revert" those mistakes.

If you have Flowise connected to a remote DB like Postgres, then you could theoretically take DB snapshots as a weak method to undo/restore.

However, ideally, it would be great if Flowise actually had some sort of native revision control built-in to the UI, so that users can easily undo/revert based on a last save.

For example, it would be awesome if you could connect Flowise to a git repo such that when a user makes a change to a chatflow, that change gets committed as a revision to the git repo.

But short-term and possibly an easier fix, would be to provide an API in Flowise to allow developers to automaticaly "export" Chatflow schemas at regular intervals.  Then, the developer could decide where to store that data (e.g., in a git repo, Google Drive, S3 bucket, or elsewhere).  An improvement upon this idea, would be to have Flowise make a remote webhook anytime a user clicks the "Save" button for a given Chatflow.  That way, rather than polling the API every day for new changes, the automation could export the Chatflow as soon as a user clicks save.

**Comment by dkindlund** [comment]
Oh, I almost forgot to include: Implement the same undo/revision tracking for Custom Tools as well.

**Comment by HenryHengZJ** [comment]
versioning is something on roadmap! for now, devs will have to write script to create a snapshot of the database for backup

**Comment by Jaredude** [comment]
Something like [Neon](https://neon.tech) Postgres handles this. Connect Flowise to a Neon Postgres and use Neon to version the DB... This allows devs to checkout versions of the Flowise DB, modify as needed without impacting Production and then merge as needed.

**Reply by HenryHengZJ** [reply]
This is a good alternative!
